Put DO WHILE into a CodeBlock and lift requirement for `default(present)` on ACC Kernels directive #346

arporter opened the pull request:
Ensures that we don't identify a DO WHILE as a NemoLoop as we don't yet support them.

rupertford commented on examples/nemo/eg3/README:
```diff
@@ -51,6 +54,9 @@ PSyclone's Schedule view of the code after adding OpenACC Kernels and
 Data directives, and 5) the transformed Fortran code with the OpenACC
 directives added.
 
+Doing the same thing with kernels_managed_mem_trans will produce similar
```
Running PSyclone with the `kernels_managed_mem_trans.py` script will ...

arporter replied:
Done.

rupertford replied:
I've updated slightly and am now happy :-)

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -4776,8 +4772,11 @@ def update(self):
         '''
-        self._add_region(start_text="KERNELS", end_text="END KERNELS",
-                         data_movement="present")
+        if self._default_present:
```
Perhaps change to ...
```
data_movement = None
if self._default_present:
    data_movement = "present"
self._add_region(start_text="KERNELS", end_text="END KERNELS",
                             data_movement=data_movement)
```

arporter replied:
Neater, thanks.

rupertford commented on src/psyclone/tests/nemo/nemo_psy_test.py:
```diff
@@ -111,6 +116,57 @@ def test_array_valued_function():
+                           api=API, line_length=False)
+    psy = PSyFactory(API, distributed_memory=False).create(invoke_info)
+    sched = psy.invokes.invoke_list[0].schedule
+    sched.view()
```
remove?

arporter replied:
Done, plus one in previous test.

rupertford commented on src/psyclone/tests/nemo/nemo_psy_test.py:
```diff
@@ -131,15 +187,13 @@ def test_multi_kern():
     loops = sched.walk(sched.children, nemo.NemoLoop)
-    sched.view()
-    gen = str(ast).lower()
     # We should have two implicit loops
```
Please update the comment.

arporter replied:
Oops, fixed.

rupertford commented on src/psyclone/tests/nemo/test_files/do_while.f90:
```diff
@@ -0,0 +1,56 @@
+  real, dimension(jpi,jpj,jpk) :: gdepw_0
+  real, dimension(jpi,jpj) :: rzisf_tbl
+  integer, dimension(jpi,jpj) :: mbkt
+  logical :: my_flag
```
I know it is only test code but shouldn't you initialise my_flag to true?

arporter replied:
For a correct Fortran program yes, but I don't initialise most of the other things, particularly the arrays. Anyway, I've added it to keep you happy :-)

rupertford replied:
Thanks. I'm happy.

rupertford commented:
Back to @arporter

arporter commented:
All comments addressed, ready for 2nd review.

arporter commented:
Realised that I'd forgotten to add the new example transformation script! I've now added it and added a line to the `check_examples` script in order to run it.

rupertford commented on examples/nemo/eg3/kernels_managed_mem_trans.py:
```diff
@@ -0,0 +1,195 @@
+# -----------------------------------------------------------------------------
```
This file is very similar to kernels_trans.py. Could you please pull out the similar utility code into a separate file.

arporter replied:
Done.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -492,6 +490,18 @@ def view(self, indent=0):
               self.ktype + "]")
 
+    @property
+    def ast(self):
```
Is this actually used anywhere? I presume so but can't find it.

arporter replied:
Yes, it's needed by the base class(es) that assume a Kernel has an `ast` property.

rupertford commented on src/psyclone/nemo.py:
```diff
@@ -492,6 +490,18 @@ def view(self, indent=0):
+                  describes this kernel.
+        :rtype: sub-class of :py:class:`fparser.two.utils.Base`
+        '''
+        return self._ast
```
nemo.py:506:1: E302 expected 2 blank lines, found 1

arporter replied:
Oops.

rupertford commented on src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py:
```diff
@@ -205,7 +205,7 @@ def test_array_syntax():
     schedule = psy.invokes.get('tra_ldf_iso').schedule
```
Not a change here but could you please check ...

C: 45, 0: third party import "from psyclone.parse.algorithm import parse" should be placed before "from fparser.common.readfortran import FortranStringReader" (wrong-import-order)
C: 46, 0: third party import "from psyclone.psyGen import PSyFactory, TransInfo, InternalError" should be placed before "from fparser.common.readfortran import FortranStringReader" (wrong-import-order)
C: 47, 0: third party import "from psyclone.transformations import TransformationError" should be placed before "from fparser.common.readfortran import FortranStringReader" (wrong-import-order)

arporter replied:
My pylint appears to be completely ambivalent about this so I've changed it :-)

rupertford commented on src/psyclone/tests/nemo/transformations/openacc/kernels_directive_test.py:
```diff
@@ -56,20 +56,6 @@
                  "end program do_loop\n")
```
Please check this ...

C: 44, 0: third party import "from psyclone.psyGen import PSyFactory, TransInfo, InternalError" should be placed before "from fparser.common.readfortran import FortranStringReader" (wrong-import-order)
C: 45, 0: third party import "from psyclone.transformations import TransformationError" should be placed before "from fparser.common.readfortran import FortranStringReader" (wrong-import-order)

arporter replied:
Again, my pylint is happy either way so I've changed it.

rupertford commented:
Nearly there. Back to @arporter.

arporter commented:
All comments addressed, ready for next review.

rupertford commented on examples/nemo/eg3/kernels_trans.py:
```diff
@@ -68,104 +68,13 @@
 
 from __future__ import print_function
 from psyclone.psyGen import TransInfo
+from kernel_utils import add_kernels
```
I know it is not modified in this PR but could you please change to `./kernels_trans.py` on line 43.

arporter replied:
Done.

rupertford commented on examples/nemo/eg3/kernel_utils.py:
```diff
@@ -0,0 +1,142 @@
+    OpenACC KERNELS directives to NEMO source. Mainly required to
+    workaround the vagaries of the PGI compiler's support for OpenACC.
+'''
+
```
I think you need `from __future__ import print_function` here.

arporter replied:
Done.

rupertford commented:
Back to @arporter

arporter commented on examples/nemo/eg3/kernel_utils.py:
```diff
@@ -0,0 +1,142 @@
+    excluded_nodes = (CodeBlock, NemoIfBlock, NemoIfClause)
+    if isinstance(node, excluded_nodes):
+        return False
+    code_blocks = node.walk(node.children, excluded_nodes)
```
Oh yes, done.

arporter commented on examples/nemo/eg3/kernel_utils.py:
```diff
@@ -0,0 +1,142 @@
+    excluded_nodes = (CodeBlock, NemoIfBlock, NemoIfClause)
+    if isinstance(node, excluded_nodes):
+        return False
+    code_blocks = node.walk(node.children, excluded_nodes)
```
I've got rid of it altogether.

arporter commented on examples/nemo/eg3/kernel_utils.py:
```diff
@@ -0,0 +1,142 @@
+        return False
+    # Check that there are no derived-type references in the sub-tree (because
+    # PGI deep-copy doesn't like them).
+    if walk_ast([node.ast], [Fortran2003.Data_Ref]):
```
As discussed, it's because we do not currently check the contents of a proposed `NemoKern` very thoroughly. This is addressed in #365. I've added a comment to the code.

arporter commented on examples/nemo/eg3/kernel_utils.py:
```diff
@@ -0,0 +1,142 @@
+    for node in nodes:
+        if isinstance(node, NemoLoop):
+            return True
+        loops = node.walk(node.children, NemoLoop)
```
Done.

arporter commented:
All suggested changes have been made. Just waiting on Travis and then it's ready for review again...
